### PR TITLE
ISSUE-1044: Add support to fetch nested fields in DruidEventMapper

### PR DESCRIPTION
**Pull request by omkreddy**

### CODE REVIEW

**Comment by omkreddy** [comment]
@arunmahadevan pinging for review

**Comment by shahsank3t** [comment]
@omkreddy Do this need any change in UI?

**Comment by omkreddy** [comment]
@Yes, we need UI change. I was thinking to raise new issue after merging this PR.

**Comment by shahsank3t** [comment]
@omkreddy Ok cool.

**Comment by omkreddy** [comment]
@shahsank3t  I will create new issue for UI change. We can  merge after this PR.

**Comment by shahsank3t** [comment]
sounds good @omkreddy

**Comment by HeartSaVioR** [comment]
Will this patch break Druid sink working with UI? If then let's file an issue for UI and handle both of them at once for not making master branch in broken state.

**Comment by omkreddy** [comment]
changes are backward compatible.  We have raised  UI issue.
https://github.com/hortonworks/streamline/issues/1046

**Comment by arunmahadevan** [comment]
Hi @priyank5485 , can you review the changes ?

File: bootstrap/components/sinks/v002__druid-sink-topology-component.json [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+      {
+        "uiName": "Block On Full",
+        "fieldName": "blockOnFull",
+        "isOptional": true,
+        "tooltip": "Whether send will block (true) or throw an exception (false) when called while the outgoing queue is full",
+        "type": "boolean",
+        "defaultValue": true
+      },
+      {
+        "uiName": "Discarded Events Stream Id",
```
Comment (priyank5485) [inline bootstrap/components/sinks/v002__druid-sink-topology-component.json]: Does this eventually translate to driud bolt's output stream? If so, not sure it is really being used given that we dont allow any connections out of a sink in SAM. Also, the edges and connections between components are created by SAM. How does SAM leverage this user input value to name the output stream in flux yaml?

File: bootstrap/components/sinks/v002__druid-sink-topology-component.json [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+                  "uiName":"fieldName",
+                  "fieldName": "fieldName",
+                  "tooltip": "fieldName",
+                  "isOptional": false,
+                  "type": "enumstring",
+                  "options": [],
+                  "hint": "inputFields"
+                },
+               {
+                 "uiName": "Parallelism",
```
Comment (priyank5485) [inline bootstrap/components/sinks/v002__druid-sink-topology-component.json]: This does not seem to be at the right level. Can you double check if it translates to the correct number of executors in the underlying storm topology? Same for v001 json.

File: streams/runners/storm/layout/src/main/java/com/hortonworks/streamline/streams/layout/storm/DruidEventMapper.java [hunk 3]
```diff
@@ -37,7 +39,60 @@ public DruidEventMapper(String eventFiledName) {
...
+        }
+
+        @Override
+        protected Map<String, Object> delegate() {
+            return event;
+        }
+
+        @Override
+        public Object get(@Nullable Object key) {
+            Object value = super.get(key);
```
Comment (priyank5485) [inline streams/runners/storm/layout/src/main/java/com/hortonworks/streamline/streams/layout/storm/DruidEventMapper.java]: Can you see if you can replace code in this method using https://github.com/hortonworks/streamline/blob/master/streams/runners/storm/runtime/src/main/java/com/hortonworks/streamline/streams/runtime/storm/StreamlineRuntimeUtil.java#L24
That way we can have everything in one place.

**Comment by priyank5485** [comment]
@omkreddy Reviewed and left some comments. I believe that main change in json is for dimensions where noNestedField is removed and hence UI will now also start passing field names with a dot convention for nesting. 

On a side note, I noticed that some druid related classes are probably in the wrong module. For example DruidBeamFactoryImpl and DruidEventMapper is in streamline-layout-storm. If thats a class thats only being used in the bolt then you can move it to the storm-runtime-storm. Thats the convention I have been following but it might not apply here.

File: bootstrap/components/sinks/v002__druid-sink-topology-component.json (hunk 1 above)
Comment (omkreddy) [reply bootstrap/components/sinks/v002__druid-sink-topology-component.json]: Agree. Currently we are just exposing the underlying druid bolt property. This is not used by SAM.  If required will remove this from druid sink.  will do in another PR.

File: bootstrap/components/sinks/v002__druid-sink-topology-component.json [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+                  "type": "enumstring",
+                  "options": [],
+                  "hint": "inputFields"
+                }
+             ]
+           }
+        ]
+      },
+      {
+        "uiName": "Parallelism",
```
Comment (omkreddy) [inline bootstrap/components/sinks/v002__druid-sink-topology-component.json]: Good catch.  moved the Parallelism field to appropriate place.

File: streams/runners/storm/layout/src/main/java/com/hortonworks/streamline/streams/layout/storm/DruidEventMapper.java [hunk 5]
```diff
@@ -37,7 +40,31 @@ public DruidEventMapper(String eventFiledName) {
...
+
+        @Override
+        protected Map<String, Object> delegate() {
+            return event;
+        }
+
+        @Override
+        public Object get(@Nullable Object key) {
+            return StreamlineEventUtils.getFieldValue(event, (String)key);
+        }
```
Comment (omkreddy) [inline streams/runners/storm/layout/src/main/java/com/hortonworks/streamline/streams/layout/storm/DruidEventMapper.java]: Thanks for the suggestion.  Actually I was looking for the existing util method.  

I moved the  StreamlineRuntimeUtils to common module and renamed as StreamlineUtils. let me know if any concerns.

**Comment by omkreddy** [comment]
@priyank5485  Thanks for the review. Updated few review comments.

Regarding moving classes to storm-runtime-storm: Agree looks like druid and opentsdb related classes are in streamline-layout-storm. Will do the cleanup after the release, to avoid any regression issues.

Since i was not moving the classes,  to access StreamlineRuntimeUtils, I moved the StreamlineRuntimeUtils to common module and renamed as StreamlineUtils. let me know if any concerns.

**Comment by priyank5485** [comment]
@omkreddy Can you file a follow up issue to clean up the changes we discussed in this PR ?Regarding renaming and moving StreamlineRuntimeUtils is it done because druid related runtime classes are in layout and it does not depend on runtime? If so, I think we should move StreamlineUtil back to runtime module since its only accessed in runtime classes. Can you make a note of that in the follow up issue? 

As far as DruidEvent is concerned do we still need that inner class? It seems it was created just to get the nested fields. If so, do you think we should remove that class completely and just use the StreamlineUtils method wherever needed? If so, you can do it in this PR or handle in the follow up issue. I will leave it up to you.

Other than the two things mentioned in this PR I am +1 once you raise a follow up issue.

**Comment by omkreddy** [comment]
@priyank5485  raised  issue here: #1070

Druid client library takes Map as input data event and key list (dimensions) to query the event. So we need pass the DruidEvent to the underlying druid library.

**Comment by arunmahadevan** [comment]
+1

**Comment by arunmahadevan** [comment]
Thanks @omkreddy for the patch. Merged to master.

**Pull request closed without merging**